pandas/tools/merge.py

```diff
@@ -981,16 +981,30 @@ def get_result(self):
             if self.axis == 0:
                 new_data = com._concat_compat([x._values for x in self.objs])
                 name = com._consensus_name_attr(self.objs)
-                return (Series(new_data, index=self.new_axes[0],
-                               name=name,
-                               dtype=new_data.dtype)
+
+                if com.is_sparse(new_data):
```

Comment:
Is this what you mean?

1. create ``core/concat.py``
2. move the logic to specify returning class and  ``core/common.py/_concat_compat`` to ``core/concat.py``
3. leave ``concat`` itself in ``tools/merge.py``